src/editor/normalize.ts

```diff
@@ -0,0 +1,135 @@
+import { isListType } from "./helpers";
+import { BlockElement, BlockElementType, SluglineElement } from "./types";
+
+const normalizeVoidSpacer = (editor: Editor, path: Path) => {
+  const prevEntry = Editor.previous(editor, { at: path });
+  const prevElem = prevEntry && (prevEntry[0] as SluglineElement);
+
+  const nextEntry = Editor.next(editor, { at: path });
+  const nextElem = nextEntry && (nextEntry[0] as SluglineElement);
+
+  // if there's a void node before or after the spacer, we don't have to do anything
+  if (
+    prevElem &&
+    Editor.isBlock(editor, prevElem) &&
+    Editor.isVoid(editor, prevElem) &&
+    prevElem.type !== BlockElementType.VoidSpacer
+  ) {
+    return;
+  }
+
+  if (
+    nextElem &&
+    Editor.isBlock(editor, nextElem) &&
+    Editor.isVoid(editor, nextElem) &&
+    nextElem.type !== BlockElementType.VoidSpacer
+  ) {
+    return;
+  }
+
+  // otherwise, remove the node
+  Transforms.removeNodes(editor, { at: path });
+};
+
+const normalizeVoidBlock = (editor: Editor, path: Path) => {
+  const pathRef = Editor.pathRef(editor, path);
+
+  const prevEntry =
+    pathRef.current &&
+    Editor.previous(editor, {
+      at: pathRef.current,
+      mode: "highest",
+    });
+
+  if (!prevEntry) {
+    Transforms.insertNodes(
+      editor,
+      {
+        type: BlockElementType.VoidSpacer,
+        children: [{ text: "" }],
+      },
+      {
+        at: [0],
+        mode: "highest",
+      }
+    );
+    return;
+  } else {
+    const [prevNode, prevPath] = prevEntry;
+    if ((prevNode as SluglineElement).type !== BlockElementType.VoidSpacer) {
+      Transforms.insertNodes(
+        editor,
+        {
+          type: BlockElementType.VoidSpacer,
+          children: [{ text: "" }],
+        },
+        { at: Editor.end(editor, prevPath), mode: "highest" }
+      );
+      return;
+    }
+  }
+
+  const nextEntry =
+    pathRef.current &&
+    Editor.next(editor, { at: pathRef.current, mode: "highest" });
+
+  if (!nextEntry) {
+    // insert a void spacer at the end of the document
+    Transforms.insertNodes(
+      editor,
+      {
+        type: BlockElementType.VoidSpacer,
+        children: [{ text: "" }],
+      },
+      {
+        at: [editor.children.length],
+      }
+    );
+    return;
+  } else {
+    const [nextNode, nextPath] = nextEntry;
+    if ((nextNode as SluglineElement).type !== BlockElementType.VoidSpacer) {
+      Transforms.insertNodes(
+        editor,
+        {
+          type: BlockElementType.VoidSpacer,
+          children: [{ text: "" }],
+        },
+        { at: Editor.before(editor, nextPath), mode: "highest" }
+      );
+      return;
+    }
+  }
+};
+
+/**
+ * Normalizes the editor, i.e., the top-level container that holds all blocks.
+ * This handles top-level layout.
+ * @param editor The editor to normalize
+ */
+export const normalizeEditor = (editor: Editor) => {
+  for (let idx = 0; idx < editor.children.length; idx++) {
```

Comment:
Suggestion: use Array.prototype.entries() for iteration, e.g.

```js
for (const [idx, node] of editor.children.entries()) {
  ...
```

Reply:
https://github.com/UWmathNEWS/slugline-web/pull/84/commits/47f43914bb06fc8a929d8d02739d83683829f688

Reply:
This actually breaks tests, oops.

Because we're actually editing the list of children as we iterate through them, `.entries()` won't work, since it gathers the index and children in one shot.

I can add a comment to this effect if you'd like.

Side note: There's a new feature in a later version of Slate that should remove the need for this function entirely, but I'll save that for another PR.

Reply:
Ahh, I didn't catch that, my bad. (sidenote: imagine if Typescript had const assertions for arguments...the joy....)